File: Backend/SER.Domain/Students/Student.cs
using CSharpFunctionalExtensions;
using SER.Domain.AdditionalQualifications;
using SER.Domain.Enterprises;
using SER.Domain.Groups;
using SER.Domain.Students.Enums;
using SER.Domain.ValueObjects;
using SER.Domain.Workplaces;
using SER.EfCore.Models.Types;
using SER.Tools.Types.IDs;
using SER.Tools.Types.Results;
using SER.Tools.Utils;

namespace SER.Domain.Students;

public class Student
{
	public ID Id { get; }
	public FullName FullName { get; } = default!;
	public Gender Gender { get; }
	public DateTime? BirthDate { get; }
	public String? PhoneNumber { get; }
	public Representative Representative { get; } = default!;
	public Boolean IsOnPaidStudy { get; }
	public String? Snils { get; }
	public Group Group { get; } = default!;
	public Passport Passport { get; } = default!;
	public WorkPlace[] WorkPlaces { get; } = default!;
	public AdditionalQualification[] AdditionalQualifications { get; } = default!;
	public TargetAgreement TargetAgreement { get; } = default!;
	public Army Army { get; } = default!;
	public SocialStatus[] SocialStatuses { get; } = default!;
	public StudentStatus Status { get; }
	public String? Address { get; }
	public Boolean IsForeignCitizen { get; }
	public String? Inn { get; }
	public String? Mail { get; }
	public String[] OtherFiles { get; } = default!;

	public WorkPlace? CurrentWorkPlace => WorkPlaces.FirstOrDefault(wp => wp.IsCurrent);
	public WorkPlace[] PreviousWorkPlaces => [.. WorkPlaces.Where(wp => !wp.IsCurrent)];

	public Student(ID id, FullName fullName, Gender gender, DateTime? birthDate, String? phoneNumber, Representative representative, Boolean isOnPaidStudy, String? snils, Group group, Passport passport, WorkPlace[] workPlaces, AdditionalQualification[] additionalQualifications, TargetAgreement targetAgreement, Army army, SocialStatus[] socialStatuses, StudentStatus status, String? address, Boolean isForeignCitizen, String? inn, String? mail, String[] otherFiles)
	{
		Id = id;
		FullName = fullName;
		Gender = gender;
		BirthDate = birthDate;
		PhoneNumber = phoneNumber;
		Representative = representative;
		IsOnPaidStudy = isOnPaidStudy;
		Snils = snils;
		Group = group;
		Passport = passport;
		WorkPlaces = workPlaces;
		AdditionalQualifications = additionalQualifications;
		TargetAgreement = targetAgreement;
		Army = army;
		SocialStatuses = socialStatuses;
		Status = status;
		Address = address;
		IsForeignCitizen = isForeignCitizen;
		Inn = inn;
		Mail = mail;
		OtherFiles = otherFiles;
	}

	public static Result<Student, Error> Create(ID? id, String? name, String? secondName, String? lastName, Gender? gender, DateTime? birthDate, String? phoneNumber, String? representativePhoneNumber, String? representativeAlias, Boolean? isOnPaidStudy, String? snils, Group? group, String? passportNumber, String? passportSeries, String? passportIssuedBy, DateTime? passportIssuedDate, String[] passportFiles, WorkPlace[] workPlaces, AdditionalQualification[] additionalQualifications, Boolean? isTargetAgreement, String? targetAgreementNumber, Enterprise? targetAgreementEnterprise, DateTime? targetAgreementDate, String[] targetAgreementFiles, Boolean? mustServeInArmy, DateTime? armyCallDate, String[] armySubpoenaFiles, SocialStatus[] socialStatuses, StudentStatus? status, String? address, Boolean? isForeignCitizen, String? inn, String? mail, String[] otherFiles)
	{
		Result<FullName, Error> createFullNameResult = FullName.Create(name, secondName, lastName);
		if(createFullNameResult.IsFailure) return createFullNameResult.Error;
		FullName fullName = createFullNameResult.Value;

		if (gender is null) return new Error("Укажите пол студента");

		if(!String.IsNullOrWhiteSpace(phoneNumber) && !Regexs.PhoneRegex.IsMatch(phoneNumber))
		{
			return new Error("Неверный формат номера телефона");
		}

		Result<Representative, Error> createRepresentativeResult = Representative.Create(representativePhoneNumber, representativeAlias);
		if(createRepresentativeResult.IsFailure) return createRepresentativeResult.Error;
		Representative representative = createRepresentativeResult.Value;

		if (isOnPaidStudy is null) return new Error("Укажите, обучается ли студент на платной основе");

		if (!String.IsNullOrWhiteSpace(snils) && !Regexs.SnilsRegex.IsMatch(snils))
		{
			return new Error("Неверно указан снилс");
		}

		if(group is null)
		{
			return new Error("Укажите группу студента");
		}

		Result<Passport, Error> createPassportResult = Passport.Create(passportNumber, passportSeries, passportIssuedBy, passportIssuedDate, passportFiles);
		if(createPassportResult.IsFailure) return createPassportResult.Error;
		Passport passport = createPassportResult.Value;

		Result<TargetAgreement, Error> createTargetAgreementResult = TargetAgreement.Create(isTargetAgreement, targetAgreementNumber, targetAgreementEnterprise, targetAgreementDate, targetAgreementFiles);
		if(createTargetAgreementResult.IsFailure) return createTargetAgreementResult.Error;
		TargetAgreement targetAgreement = createTargetAgreementResult.Value;

		Result<Army, Error> createArmyResult = Army.Create(mustServeInArmy, armyCallDate, armySubpoenaFiles);
		if(createArmyResult.IsFailure) return createArmyResult.Error;
		Army army = createArmyResult.Value;

		if (status is null) return new Error("Укажите статус студента");

		if (isForeignCitizen is null) return new Error("Укажите, является ли студент иностранным гражданином");

		if (!String.IsNullOrWhiteSpace(mail) && !Regexs.MailRegex.IsMatch(mail))
		{
			return new Error("Неверно указана электронная почта");
		}

		if (!String.IsNullOrWhiteSpace(inn) && !Regexs.HumanInnRegex.IsMatch(inn))
		{
			return new Error("Неверный формат ИНН");
		}

		ID _id = id ?? ID.New();

		return new Student(_id, fullName, gender.Value, birthDate, phoneNumber, representative, isOnPaidStudy.Value, snils, group, passport, workPlaces, additionalQualifications, targetAgreement, army, socialStatuses, status.Value, address, isForeignCitizen.Value, inn, mail, otherFiles);
	}
}

File: Frontend/students-employment-records-site/src/domain/students/models/student.ts
import { AdditionalQualification } from "@/domain/additionalQualifications/models/additionalQualification"
import { Enterprise } from "@/domain/enterprises/models/enterprise"
import { Group } from "@/domain/groups/models/group"
import { Workplace } from "@/domain/workplaces/models/workplace"
import { WorkplaceBlank } from "@/domain/workplaces/models/workplaceBlank"
import { Gender } from "../enums/gender"
import { SocialStatus } from "../enums/socialStatus"
import { StudentStatus } from "../enums/studentStatus"
import { StudentBlank } from "./studentBlank"

export class Student {
    constructor(
        public readonly id: string,
        public readonly name: string,
        public readonly secondName: string,
        public readonly lastName: string | null,
        public readonly gender: Gender,
        public readonly birthDate: Date | null,
        public readonly phoneNumber: string | null,
        public readonly representativePhoneNumber: string | null,
        public readonly representativeAlias: string | null,
        public readonly isOnPaidStudy: boolean,
        public readonly snils: string | null,
        public readonly group: Group,
        public readonly passportNumber: string | null,
        public readonly passportSeries: string | null,
        public readonly passportIssuedBy: string | null,
        public readonly passportIssuedDate: Date | null,
        public readonly passportFiles: string[],
        public readonly prevWorkplaces: Workplace[],
        public readonly currentWorkplace: Workplace | null,
        public readonly additionalQualifications: AdditionalQualification[],
        public readonly isTargetAgreement: boolean,
        public readonly targetAgreementNumber: string | null,
        public readonly targetAgreementFiles: string[],
        public readonly targetAgreementDate: Date | null,
        public readonly targetAgreementEnterprise: Enterprise | null,
        public readonly mustServeInArmy: boolean,
        public readonly armySubpoenaFiles: string[],
        public readonly armyCallDate: Date | null,
        public readonly socialStatuses: SocialStatus[],
        public readonly status: StudentStatus,
        public readonly address: string | null,
        public readonly isForeignCitizen: boolean,
        public readonly inn: string | null,
        public readonly mail: string | null,
        public readonly otherFiles: string[]
    ) {}

    public get displayName(): string {
        return `${this.secondName} ${this.name} ${this.lastName ?? ""}`
    }

    public toBlank(): StudentBlank {
        const currentWorkplace = this.currentWorkplace
            ? WorkplaceBlank.create(this.currentWorkplace)
            : null
        const prevWorkplaces = this.prevWorkplaces.map(WorkplaceBlank.create)
        return {
            id: this.id,
            name: this.name,
            secondName: this.secondName,
            lastName: this.lastName,
            gender: this.gender,
            birthDate: this.birthDate,
            phoneNumber: this.phoneNumber,
            representativePhoneNumber: this.representativePhoneNumber,
            representativeAlias: this.representativeAlias,
            isOnPaidStudy: this.isOnPaidStudy,
            snils: this.snils,
            group: this.group,
            passportNumber: this.passportNumber,
            passportSeries: this.passportSeries,
            passportIssuedBy: this.passportIssuedBy,
            passportIssuedDate: this.passportIssuedDate,
            passportFiles: this.passportFiles,
            prevWorkplaces,
            currentWorkplace,
            additionalQualifications: this.additionalQualifications,
            isTargetAgreement: this.isTargetAgreement,
            targetAgreementNumber: this.targetAgreementNumber,
            targetAgreementFiles: this.targetAgreementFiles,
            targetAgreementDate: this.targetAgreementDate,
            targetAgreementEnterprise: this.targetAgreementEnterprise,
            mustServeInArmy: this.mustServeInArmy,
            armySubpoenaFiles: this.armySubpoenaFiles,
            armyCallDate: this.armyCallDate,
            socialStatuses: this.socialStatuses,
            status: this.status,
            address: this.address,
            isForeignCitizen: this.isForeignCitizen,
            inn: this.inn,
            mail: this.mail,
            otherFiles: this.otherFiles,
        }
    }

    public static fromAny(any: any): Student {
        const birthDate = any.birthDate ? new Date(any.birthDate) : null
        const group = Group.fromAny(any.group)
        const passportIssuedDate = any.passportIssuedDate ? new Date(any.passportIssuedDate) : null
        const prevWorkplaces = (any.prevWorkplaces as any[]).map(Workplace.fromAny)
        const currentWorkplace = any.currentWorkplace
            ? Workplace.fromAny(any.currentWorkplace)
            : null
        const additionalQualifications = (any.additionalQualifications as any[]).map(
            AdditionalQualification.fromAny
        )
        const targetAgreementDate = any.targetAgreementDate
            ? new Date(any.targetAgreementDate)
            : null
        const targetAgreementEnterprise = any.targetAgreementEnterprise
            ? Enterprise.fromAny(any.targetAgreementEnterprise)
            : null
        const armyCallDate = any.armyCallDate ? new Date(any.armyCallDate) : null

        return new Student(
            any.id,
            any.name,
            any.secondName,
            any.lastName,
            any.gender,
            birthDate,
            any.phoneNumber,
            any.representativePhoneNumber,
            any.representativeAlias,
            any.isOnPaidStudy,
            any.snils,
            group,
            any.passportNumber,
            any.passportSeries,
            any.passportIssuedBy,
            passportIssuedDate,
            any.passportFiles,
            prevWorkplaces,
            currentWorkplace,
            additionalQualifications,
            any.isTargetAgreement,
            any.targetAgreementNumber,
            any.targetAgreementFiles,
            targetAgreementDate,
            targetAgreementEnterprise,
            any.mustServeInArmy,
            any.armySubpoenaFiles,
            armyCallDate,
            any.socialStatuses,
            any.status,
            any.address,
            any.isForeignCitizen,
            any.inn,
            any.mail,
            any.otherFiles
        )
    }
}
